src/features/runtime-checks.js

```diff
@@ -0,0 +1,249 @@
+        }
+        this.#listeners.push([...args])
+    }
+
+    removeEventListener (...args) {
+        const el = this.getElement()
+        if (el) {
+            return el.removeEventListener(...args)
+        }
+        this.#listeners = this.#listeners.filter((listener) => {
+            return listener[0] !== args[0] || listener[1] !== args[1]
+        })
+    }
+
+    toString () {
+        // TODO
+        let interfaceName = ''
+        if (this._tagName === 'script') {
+            interfaceName = 'Script'
+        }
+        return `[object HTML${interfaceName}Element]`
+    }
+
+    get tagName () {
+        return this.#tagName.toUpperCase()
+    }
+
+    get nodeName () {
+        return this.tagName
+    }
+
+    remove () {
+        const el = this.getElement()
+        if (el) {
+            return el.remove()
+        }
+        return super.remove()
+    }
+
+    removeChild (child) {
+        const el = this.getElement()
+        if (el) {
+            return el.removeChild(child)
+        }
+        return super.removeChild(child)
+    }
+}
+
+/**
+ * Overrides the instanceof checks to make the custom element interface pass an instanceof check
+ * @param {Object} elementInterface
+ */
+function overloadInstanceOfChecks (elementInterface) {
+    const proxy = new Proxy(elementInterface[Symbol.hasInstance], {
+        apply (fn, scope, args) {
+            if (args[0] instanceof DDGRuntimeChecks) {
+                return true
+            }
+            return Reflect.apply(fn, scope, args)
+        }
+    })
+    Object.defineProperty(elementInterface, Symbol.hasInstance, {
+        value: proxy
+    })
+}
+
+/**
+ * Returns true if the tag should be intercepted
+ * @param {string} tagName
+ * @returns {boolean}
+ */
+function shouldInterrogate (tagName) {
+    const interestingTags = ['script']
+    if (!interestingTags.includes(tagName)) {
+        return false
+    }
+    const stack = getStack()
+    const scriptOrigins = [...getStackTraceOrigins(stack)]
+    const isInterestingHost = scriptOrigins.some(origin => {
+        return interestingHosts.find(hostname => matchHostname(origin, hostname))
+    })
+    return isInterestingHost
+}
+
+function overrideCreateElement () {
+    const proxy = new DDGProxy(featureName, document, 'createElement', {
+        apply (fn, scope, args) {
+            if (args.length >= 1) {
+                const initialTagName = args[0].toLowerCase()
+                if (shouldInterrogate(initialTagName)) {
+                    args[0] = 'ddg-runtime-checks'
+                    const el = Reflect.apply(fn, scope, args)
+                    el.setTagName(initialTagName)
+                    return el
+                }
+            }
+            return Reflect.apply(fn, scope, args)
+        }
+    })
+    proxy.overload()
+    initialCreateElement = proxy._native
+}
+
+export function load () {
+    customElements.define('ddg-runtime-checks', DDGRuntimeChecks)
```

Comment:
This [can throw](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define), for example if the extension reloads and reruns this code, or if there is a conflict with existing custom elements in the page. Should we catch and handle that error here?

Reply:
We don't have any idempotency guarantees for C-S-S (in fact in Apple code we have to do a grubby little hack to prevent it from triggering twice).

Additionally the page shouldn't run first (as you well know this isn't always the case); so yeah I'll catch this but it "shouldn't ever happen™️"